### PR TITLE
Add support for postprocess exported model for block scale swizzling and support for different padding strategy

### PR DESCRIPTION
**Pull request by ynankani**
### What does this PR do?

Type of change: ?  new feature

Adds post-processing support for exported diffusion model checkpoints to enable NVFP4 block scale swizzling and configurable padding strategies. This allows exported quantized checkpoints to be directly consumed by inference runtimes (e.g., ComfyUI with comfy_kitchen) that require cuBLAS 2-D block-scaling-factors layout.
Changes:
1) Unified post-processing step (_postprocess_safetensors): Loads saved safetensors files and applies merge, padding, swizzle, and quantization metadata injection in a single pass. 
2) NVFP4 scale swizzle (swizzle_nvfp4_scales): Rearranges block scales from ModelOpt's flat [rows, cols // 16] layout to cuBLAS 2-D tiled layout per the cuBLAS specification.
3) Configurable padding (pad_nvfp4_weights): Pads NVFP4 weight and scale tensors to multiples of 16, with "row" (rows only) or "row_col" (both dimensions) strategies.
4) Standalone quantization metadata (build_layerwise_quant_metadata): Extracted from merge_diffusion_checkpoint so _quantization_metadata can be injected independently of merging — works for both merged (LTX-2) and standalone (Flux2) exports.
5) Bug fix (conversion.py): Wrapped yield in try/finally in set_quantizer_by_cfg_context so quantizer states are always restored, fixing an issue when yield fails.
### Usage

```python
# LTX-2 export with merge + swizzle + padding
export_hf_checkpoint(
  pipeline,
  export_dir="./output",
  merged_base_safetensor_path="./ltx-2-22b-dev.safetensors",
  enable_swizzle_layout=True,
  padding_strategy="row_col",
  enable_layerwise_quant_metadata=True,
)
# Flux2 standalone export with swizzle + padding (no merge needed)
export_hf_checkpoint(
  transformer,
  export_dir="./output",
  enable_swizzle_layout=True,
  padding_strategy="row_col",
)
# Via quantize.py CLI
python quantize.py \
  --model ltx-2 --format fp4 \
  --extra-param merged_base_safetensor_path=./ltx-2-22b-dev.safetensors \
  --extra-param enable_swizzle_layout=true \
  --extra-param padding_strategy=row_col \
  --hf-ckpt-dir ./output
```

### Testing
1) Exported LTX-2.3 NVFP4 with swizzle + padding + merged base checkpoint. Verified checkpoint has correct uint8 weights, float8_e4m3fn scales in swizzled layout, and _quantization_metadata . Ran the checkpoint with ComfyUI
2) Exported Flux2 NVFP4 with swizzle + padding. Verified checkpoint has correct uint8 weights, float8_e4m3fn scales in swizzled layout, and _quantization_metadata . Ran the checkpoint with ComfyUI

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?: ✅ 
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`:  N/A 
- Did you write any new necessary tests?: N/A 
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?:  N/A 

## Summary by CodeRabbit

* **New Features**
  * Diffusers export: optional NVFP4 support — swizzle layout, row/row_col padding, and optional per-layer quantization metadata; exports are now post-processed to apply these options.
  * Export flow accepts new flags to enable swizzle, padding strategy, and layerwise metadata.

* **Bug Fixes**
  * Quantizer context manager now always restores state, including on exceptions.

* **Tests**
  * Added unit tests for NVFP4 padding, swizzling, metadata injection, and post-processing.

* **Documentation**
  * README example updated to show swizzle and padding flags.

### CODE REVIEW

File: modelopt/torch/export/unified_export_hf.py [hunk 1]
```diff
@@ -28,7 +28,9 @@
 
 import torch
 import torch.nn as nn
```
Comment (cjluo-nv) [inline modelopt/torch/export/unified_export_hf.py:30]: `load_file` is added here but `safe_open` (already available from safetensors) would be better for reading metadata at line ~196, avoiding the manual `struct.unpack` header parsing.

File: modelopt/torch/export/unified_export_hf.py [hunk 2]
```diff
@@ -121,48 +123,111 @@ def _is_enabled_quantizer(quantizer):
...
+        model_type: Key into ``DIFFUSION_MERGE_FUNCTIONS`` (e.g., ``"ltx2"``).
+        hf_quant_config: Quantization config dict to embed in metadata.
+        enable_layerwise_quant_metadata: Whether to build per-layer metadata.
+        padding_strategy: ``"row"``, ``"row_col"``, or None.
+        enable_swizzle_layout: Whether to swizzle block scales.
+    """
+    import struct
+
+    safetensor_files = sorted(export_dir.glob("*.safetensors"))
+    if not safetensor_files:
```
Comment (cjluo-nv) [inline modelopt/torch/export/unified_export_hf.py:206]: Move `import struct` to the top of the file with other stdlib imports. Inline imports are typically reserved for optional/heavy dependencies, not stdlib modules.

File: modelopt/torch/export/unified_export_hf.py [hunk 3]
```diff
@@ -121,48 +123,111 @@ def _is_enabled_quantizer(quantizer):
...
+
+    if list(export_dir.glob("*.safetensors.index.json")) and (
+        merged_base_safetensor_path is not None or enable_layerwise_quant_metadata
+    ):
+        raise NotImplementedError(
+            "Post-processing sharded safetensors is not supported. "
+            "Export with a larger max_shard_size or disable merge/metadata options."
+        )
+
+    for sf_path in safetensor_files:
```
Comment (cjluo-nv) [inline modelopt/torch/export/unified_export_hf.py:223]: Consider replacing this manual header parsing with `safetensors.safe_open`:
```python
with safe_open(str(sf_path), framework="pt") as f:
metadata = dict(f.metadata() or {})
```
This is the pattern used in `_merge_ltx2` and is less fragile than binary header parsing.

File: modelopt/torch/export/unified_export_hf.py [hunk 4]
```diff
@@ -121,48 +123,111 @@ def _is_enabled_quantizer(quantizer):
...
+            header_size = struct.unpack("<Q", f.read(8))[0]
+            header = json.loads(f.read(header_size))
+        metadata = header.get("__metadata__", None) or {}
+
+        # Clone tensors so the memory-mapped file handle from load_file is
+        # released before we overwrite the same path (required on Windows).
+        sd = {k: v.clone() for k, v in sd.items()}
+
+        if merged_base_safetensor_path is not None and model_type is not None:
+            sd, base_metadata = merge_diffusion_checkpoint(
```
Comment (cjluo-nv) [inline modelopt/torch/export/unified_export_hf.py:229]: `metadata.update(base_metadata)` reverses the old merge precedence. Previously export keys (`_export_format`, `_class_name`) always won. Now base checkpoint keys would win if they overlap. Consider using: `base_metadata.update(metadata); metadata = base_metadata` to preserve the old semantics.

File: modelopt/torch/export/diffusers_utils.py [hunk 5]
```diff
@@ -797,23 +830,119 @@ def merge_diffusion_checkpoint(
     if hf_quant_config is not None:
```
Comment (cjluo-nv) [inline modelopt/torch/export/diffusers_utils.py:852]: This `quantization_config` write is now redundant — `_postprocess_safetensors` always writes it again when `hf_quant_config is not None`. Consider removing this line (or passing `hf_quant_config=None` from the caller) to avoid confusion about where metadata is authored.

**Review by cjluo-nv — COMMENTED** [review]
**Summary**: Adds NVFP4 post-processing pipeline for exported diffusion model checkpoints — padding, cuBLAS block-scale swizzling, and independent metadata injection — refactored out of the save step into a separate `_postprocess_safetensors` pass. Also fixes a context-manager restoration bug in `set_quantizer_by_cfg_context`.

**Issues Found**:

1. **[Correctness]** Metadata merge order is reversed from prior behavior. Old code: `base_metadata.update(export_metadata)` (export keys like `_export_format` win). New code in `_postprocess_safetensors`: `metadata.update(base_metadata)` (base keys win). Currently benign because key sets likely don't overlap, but fragile if base checkpoints ever contain `_export_format` or `_class_name`.
   - File: `modelopt/torch/export/unified_export_hf.py`, new lines ~207-210

2. **[Correctness]** `quantization_config` is written twice when merge is active: once inside `merge_diffusion_checkpoint` (line 830 of diffusers_utils.py) and again explicitly in `_postprocess_safetensors` (line ~220). The second write is identical, so not a bug, but the redundancy means the merge function's metadata write is now dead code for merged paths.
   - File: `modelopt/torch/export/unified_export_hf.py`, lines ~218-220 and `diffusers_utils.py` line 830

3. **[Correctness]** `_postprocess_safetensors` is now invoked for **all** quantized components, including those saved via `save_pretrained`. Previously these components got no merge/swizzle/padding/metadata in the safetensors file. This is likely intentional for extending NVFP4 support, but it's a behavioral change worth confirming — especially for components that aren't the main transformer (e.g., a quantized VAE could have its scales swizzled too).
   - File: `modelopt/torch/export/unified_export_hf.py`, lines ~1047-1055

4. **[Readability]** `import struct` is inside the `_postprocess_safetensors` function body. It's a stdlib module — move it to the module-level imports for consistency with the rest of the file.
   - File: `modelopt/torch/export/unified_export_hf.py`, line ~186

5. **[Readability]** Metadata is read via manual binary header parsing (`struct.unpack` + `json.loads`). The `safetensors` library already provides `safe_open(...).metadata()` for this purpose (used elsewhere in this codebase, e.g., `_merge_ltx2`). Using `safe_open` would be simpler and less fragile.
   - File: `modelopt/torch/export/unified_export_hf.py`, lines ~194-197

6. **[Tests]** The `try/finally` fix in `conversion.py` (the most safety-critical change in this PR) has **zero test coverage**. Codecov confirms 12 missing lines. A test that raises inside the `set_quantizer_by_cfg_context` body and verifies quantizer state is still restored would validate this fix.

7. **[Tests]** No test covers the `save_pretrained` → `_postprocess_safetensors` path (issue #3 above). The existing tests only exercise the custom-save path via standalone safetensors files.

8. **[Correctness]** `_to_blocked` internally pads scale rows to multiples of 128 and **changes the output tensor shape** (e.g., input `(16, 4)` → output `(128, 4)`). This means after swizzle, `weight.shape[0] != weight_scale.shape[0]`. The test explicitly asserts this, and cuBLAS likely requires it, but this shape mismatch should be documented in the `swizzle_nvfp4_scales` docstring so future maintainers don't "fix" it.

**Suggestions**:
- Consider passing `hf_quant_config=None` to `merge_diffusion_checkpoint` in `_postprocess_safetensors` to avoid the redundant `quantization_config` write, since it's set explicitly afterward anyway.
- `_find_nvfp4_layers` is called separately by `pad_nvfp4_weights` and `swizzle_nvfp4_scales`. When both are invoked in `_postprocess_safetensors`, detection runs twice over the same state dict. A minor optimization would be to compute it once and pass it in, but not blocking.
- The copyright header in `test_nvfp4_utils.py` says 2024 — should be 2026.

**Overall Assessment**: The core functionality (padding, swizzle, metadata) is well-structured and correctly extracted. The `conversion.py` try/finally fix is a genuine bug fix. However, the metadata precedence reversal and the behavioral change for `save_pretrained` components need explicit confirmation, and the try/finally fix needs a test.

File: modelopt/torch/export/diffusers_utils.py [hunk 6]
```diff
@@ -797,23 +830,119 @@ def merge_diffusion_checkpoint(
...
+        padded_cols = n_col_blocks * 4
+        padded = input_matrix
+        if (rows, cols) != (padded_rows, padded_cols):
+            padded = torch.zeros(
+                (padded_rows, padded_cols),
+                device=input_matrix.device,
+                dtype=input_matrix.dtype,
+            )
+            padded[:rows, :cols] = input_matrix
+        blocks = padded.view(n_row_blocks, 128, n_col_blocks, 4).permute(0, 2, 1, 3)
```
Comment (cjluo-nv) [inline modelopt/torch/export/diffusers_utils.py:959]: The docstring should note that the output shape may differ from the input shape (rows padded to multiples of 128 for cuBLAS alignment). This is a non-obvious permanent transformation that could surprise maintainers.

File: tests/unit/torch/export/test_nvfp4_utils.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (cjluo-nv) [inline tests/unit/torch/export/test_nvfp4_utils.py]: Copyright year says 2024 — should be 2026 to match the current year.

File: modelopt/torch/export/unified_export_hf.py [hunk 8]
```diff
@@ -971,14 +1044,20 @@ def _export_diffusers_checkpoint(
...
-                        component_export_dir,
-                        merged_base_safetensor_path,
-                        hf_quant_config,
-                        model_type,
-                    )
-            # Step 7: Update config.json with quantization info
+                    _save_component_state_dict_safetensors(component, component_export_dir)
+
+            # Step 7: Post-process — merge, metadata, padding, swizzle
+            _postprocess_safetensors(
```
Comment (cjluo-nv) [inline modelopt/torch/export/unified_export_hf.py:1081]: This `_postprocess_safetensors` call now runs for `save_pretrained` components too (the `if hasattr(component, "save_pretrained")` branch above). In the old code, those components got no merge/swizzle/metadata. Please confirm this behavioral change is intentional and add a test for it.

Comment (ynankani) [reply modelopt/torch/export/unified_export_hf.py:1081]: For diffusers based model like flux2 the save_pretrained branch will hit and the swizzling, metadata will be required for those models as well.

File: modelopt/torch/export/unified_export_hf.py [hunk 9]
```diff
@@ -1162,15 +1229,31 @@ def export_hf_checkpoint(
             to export. If None, all quantized components are exported.
         extra_state_dict: Extra state dictionary to add to the exported model.
         max_shard_size: Maximum size of each safetensors shard file. Defaults to "10GB".
-        **kwargs: Internal-only keyword arguments. Supported key: merged_base_safetensor_path
-            (str, optional). When provided, merges the exported diffusion transformer
-            weights with non-transformer components (VAE, vocoder, text encoders, etc.)
-            from this base safetensors file to produce a single-file checkpoint
-            compatible with ComfyUI. Value should be the path to a full base model
-            ``.safetensors`` file (e.g. ``"path/to/ltx-2-19b-dev.safetensors"``).
+        **kwargs: Internal-only keyword arguments. Supported keys:
```
Comment (jingyu-ml) [inline modelopt/torch/export/unified_export_hf.py]: Instead of extracting these ComfyUI-specific params here, could we pass **kwargs through to _export_diffusers_checkpoint and have it call a postprocess_for_comfyui function that owns the extraction/validation? That keeps export_hf_checkpoint runtime-agnostic and puts all ComfyUI-specific logic in one place.

Comment (jingyu-ml) [reply modelopt/torch/export/unified_export_hf.py]: The ComfyUI-specific parameter descriptions (merged_base_safetensor_path, enable_swizzle_layout, padding_strategy, enable_layerwise_quant_metadata) don't need to be documented here. Just pass **kwargs through to _export_diffusers_checkpoint -> postprocess_for_comfyui, and document them in postprocess_for_comfyui's docstring instead.

**Comment by jingyu-ml** [comment]
Have you tested the checkpoint export and tried deploying it on ComfyUI with NVFP4?

**Comment by ynankani** [comment]
> Have you tested the checkpoint export and tried deploying it on ComfyUI with NVFP4?

Yes, I’ve tested the LTX2.3 and Flux2 models exported with ModelOpt for inference in ComfyUI using NVFP4. During the process, I had to make adjustments for nibble order handling and name mapping within ComfyUI. These required modifications have been shared with the ComfyUI team, and I’ll retest once their updates are implemented.

**Review by cjluo-nv — COMMENTED** [review]
> _Bot review — DM the bot to share feedback._

Most prior reviewer concerns have been addressed (extra_params validation, sharded-checkpoint guard, `safe_open` for metadata, scale-only padding test, swizzle-shape assertion in tests, copyright year). Two unresolved items warrant a human look before approval:

- :speech_balloon: Prior review (cjluo-nv) flagged that the `try/finally` fix in `modelopt/torch/quantization/conversion.py` — the most safety-critical change in this PR, and one that affects every caller of `set_quantizer_by_cfg_context` (LLM PTQ, vLLM fakequant, etc., not just the new diffusion path) — has **zero test coverage**. Author has not responded and no test was added in subsequent commits. A small unit test that raises inside the context body and asserts quantizer state is still restored would validate this fix and prevent regressions in unrelated callers.
- :speech_balloon: Prior review (jingyu-ml) suggested moving the ComfyUI-specific kwargs (`merged_base_safetensor_path`, `enable_swizzle_layout`, `padding_strategy`, `enable_layerwise_quant_metadata`) out of `export_hf_checkpoint`'s public signature and into a `postprocess_for_comfyui` helper so `export_hf_checkpoint` stays runtime-agnostic. The author replied to the deployment-test question but not to this design point. The current `**kwargs` plumbing leaves these public-API knobs undocumented at the entry point — worth a maintainer's call.

Minor (non-blocking): `padded_count` in `pad_nvfp4_weights` is still computed but never used or returned (CodeRabbit flagged this; not addressed).

Design-review check: this is an incremental extension of the existing diffusers export pipeline (no new subsystem/DSL/registry), so the architectural protocol doesn't materially change the verdict.

**Review by jingyu-ml — APPROVED** [review]
LGTM. Please address the bot’s minor concern before merging the MR.

**Comment by jingyu-ml** [comment]
cc @Edwardf0t1

File: modelopt/torch/export/unified_export_hf.py (hunk 9 above)
Comment (ynankani) [reply modelopt/torch/export/unified_export_hf.py]: Moved all the extraction to _postprocess_safetensors so that all comfy specific logic in one place and documented these params in this function
